### PR TITLE
[MM-42194] Get file information from a deleted post

### PR DESCRIPTION
**Pull request by santoniriccardo**
#### Summary

Adds a query parameter include_deleted allowing an admin to retrieve contents of post whether or not it is deleted.

If the query parameter is not present, maintains current behavior.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/20231

https://mattermost.atlassian.net/browse/MM-42194
#### Release Note

```release-note
Adds query parameter 'include_deleted' to endpoint: {{[http://your-mattermost-url.com/api/v4/posts/{post_id}/files/info}}](http://your-mattermost-url.com/api/v4/posts/%7Bpost_id%7D/files/info%7D%7D)
```

### CODE REVIEW

File: model/client4.go [hunk 1]
```diff
@@ -4355,6 +4355,24 @@ func (c *Client4) GetFileInfosForPost(postId string, etag string) ([]*FileInfo,
...
+	var list []*FileInfo
+	if r.StatusCode == http.StatusNotModified {
+		return list, BuildResponse(r), nil
+	}
+	if jsonErr := json.NewDecoder(r.Body).Decode(&list); jsonErr != nil {
+		return nil, nil, NewAppError("GetFileInfosForPostIncludeDeleted", "api.unmarshal_error", nil, jsonErr.Error(), http.StatusInternalServerError)
+	}
+	return list, BuildResponse(r), nil
+}
+
```
Comment (santoniriccardo) [inline model/client4.go:4375]: Note, decided to implement completely seperate client function for the specific case for including deleted as this was the pattern that I noticed throughout the file ([ex](https://github.com/mattermost/mattermost-server/blob/54a88fb53268874aad436c77c72586ac24eb3e70/model/client4.go#L1919-L1950)), as well as the fact that changing the current function would deprecate current users of the library.

That being said, not sure that this is the best pattern for scalability down the line, especially if more options are anticipated.

**Comment by santoniriccardo** [comment]
@amyblais Tagging you as you originally assigned me. Looking for relevant reviewers for this PR.

File: app/post.go [hunk 2]
```diff
@@ -1182,6 +1182,11 @@ func (a *App) DeletePost(postID, deleteByID string) (*model.Post, *model.AppErro
 
 	a.invalidateCacheForChannelPosts(post.ChannelId)
 
+	if post.FileIds != nil {
+		a.Srv().Store.FileInfo().InvalidateFileInfosForPostCache(postID, true)
+		a.Srv().Store.FileInfo().InvalidateFileInfosForPostCache(postID, false)
+	}
+
```
Comment (ashishbhate) [inline app/post.go:1189]: Could we just do this right after line 1177?

**Review by ashishbhate — APPROVED** [review]
LGTM

**Comment by ashishbhate** [comment]
@Willyfrog added you as an optional reviewer since I will be away for a couple weeks.

**Comment by Willyfrog** [comment]
@santoniriccardo let us know when it's ready for review, as I see you are still making changes :)

**Comment by santoniriccardo** [comment]
@Willyfrog Good to go, just added in the recent suggestion from @ashishbhate.

**Review by Willyfrog — APPROVED** [review]

**Review by vish9812 — APPROVED** [review]

**Comment by jgilliam17** [comment]
Thanks @santoniriccardo 
Tested, looks good to merge - pending E2E report - updating branch again - cycle is failing to complete
- Verified admin can get information from a deleted post

**Review by jgilliam17 — APPROVED** [review]
Thanks for your patience @santoniriccardo 
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/221187-d1a344e-onprem-ent-server-pr-20279-mattermost/mm-ee-testd1a344e/mochawesome.html) - no PR related failures.

**Pull request merged**